### ISSUE TITLE
[linux] Dangling images with Docker 23

### ISSUE THREAD

**Issue by stasadev** [ddev/ddev#4871]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Expected Behavior

I have the same problem as in https://github.com/ddev/ddev/issues/1912, but my Docker creates dangling images on start for `ddev-ssh-agent`, `ddev-webserver`, `ddev-dbserver`. On DDEV stop/start, a new set of dangling images is created.

```
$ docker images -f dangling=true
REPOSITORY   TAG       IMAGE ID       CREATED          SIZE
<none>       <none>    bcb482c0e876   5 minutes ago   1.2GB
<none>       <none>    f62dc0503c79   5 minutes ago   798.5MB
<none>       <none>    9656b65d5ef5   5 minutes ago   119MB
```

Docker builder should not create dangling images `<none>:<none>`.

### Actual Behavior

Dangling images are created on Linux.

I don't know if other platforms (Windows, macOS) are affected, as [v23](https://github.com/moby/moby/releases/tag/v23.0.0) release notes say "Set Buildx and BuildKit as the default builder on Linux".

There are differences in how BuildKit and the legacy builder handle multi-stage builds. For more information, see https://docs.docker.com/build/building/multi-stage/

### Steps To Reproduce

1. Use Docker 23 on Linux with [Buildx](https://github.com/docker/buildx) and [BuiltKit](https://github.com/moby/buildkit)
3. Make sure there are no dangling images `docker rmi $(docker images -f dangling=true -q)`
4. Run `ddev start <project>`
5. Check for `docker images -f dangling=true`

### Related issue:

- https://github.com/ddev/ddev/issues/1912

I'm not sure if this is what DDEV needs to handle, because BuildKit is still a work in progress https://github.com/moby/moby/issues/40379.

### Alternatives:

- Use Docker 20.10.x
- Set `DOCKER_BUILDKIT=0` for the legacy builder - I used this approach.

**Comment by rfay** [comment]
Wouldn't this be something to file upstream? In moby/moby for example?

Note that @rpkoller has been studying dangling images for some time and wondering where they come from.

**Comment by stasadev** [comment]
This issue is for future DDEV users who may come across this.

> Wouldn't this be something to file upstream? In moby/moby for example?

Yes, probably. I just don't know yet how to create a minimal example Dockerfile with this bug.

**Comment by rfay** [comment]
The original images do require buildx (which I think means using BUILDKIT) because they're multi-arch images. 

However, the extra layer(s) added by custom Dockerfile and by a couple of other layers that get added at first `start` do *not* require this. So maybe that is the issue. 

Most of the extra layers are done in https://github.com/ddev/ddev/blob/8f4e549971df76821c8612c64940d2f2ac02fbdb/pkg/ddevapp/config.go#L889-L981

Are you able to demonstrate this problem with a trivial project (like `ddev debug test` creates)? Or does it only happen if you have extra content like a Dockerfile, etc?

**Comment by stasadev** [comment]
> Are you able to demonstrate this problem with a trivial project (like ddev debug test creates)?

Yes.

> Or does it only happen if you have extra content like a Dockerfile, etc?

I mean, I need to show something reproducible, like Dockerfile, to Docker devs.

```
$ docker images -f dangling=true
REPOSITORY   TAG       IMAGE ID   CREATED   SIZE
```

Run [ddev debug test](https://gist.github.com/stasadev/0b719f6ee1159f45351d0e93f67bec22).

Then I see a dangling image for `ddev-ssh-agent`:

```
$ docker images -f dangling=true
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
<none>       <none>    1444ab586442   2 months ago   119MB
```

Then I restart this test project:

```
$ ddev restart tryddevproject-5811
Restarting project tryddevproject-5811... 
Container ddev-tryddevproject-5811-db  Removed 
Container ddev-tryddevproject-5811-web  Removed 
Network ddev-tryddevproject-5811_default  Removed 
Network ddev-tryddevproject-5811_default  Created 
Container ddev-tryddevproject-5811-web  Started 
Container ddev-tryddevproject-5811-db  Started 
Container ddev-router  Started 
Restarted tryddevproject-5811 
Your project can be reached at https://tryddevproject-5811.ddev.site https://127.0.0.1:32863
```

And see one more dangling image for `ddev-webserver`:

```
$ docker images -f dangling=true
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
<none>       <none>    611c01362fba   2 months ago   1.16GB
<none>       <none>    1444ab586442   2 months ago   119MB
```

**Comment by rfay** [comment]
It seems it's leaving a copy ddev-webserver there as dangling (at least I assume that since it says 2 months ago...)

This may also have to do with the build stanza itself, https://github.com/ddev/ddev/blob/765aebebfcfeeb5dc2b0aefd474e31f3ff3e0ed1/pkg/ddevapp/app_compose_template.yaml#L87-L95

Maybe there's something not quite right there. Like the `image` statement... is it even useful?

**Comment by stasadev** [comment]
I have the project still running, trying to inspect `611c01362fba`

```
$ docker images -f dangling=true
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
<none>       <none>    611c01362fba   2 months ago   1.16GB
<none>       <none>    1444ab586442   2 months ago   119MB
```

```
$ docker images | grep 1.16GB
<none>                                    <none>                                     d3c8a3de8610   2 months ago     1.16GB
drud/ddev-webserver                       v1.21.5-tryddevproject-5811-built          611c01362fba   2 months ago     1.16GB
drud/ddev-webserver                       v1.21.5                                    bcd7a520efce   2 months ago     1.16GB
```

```
$ docker container inspect $(docker container ls -aq) --format "Image: {{.Image}} Container: {{.Id}}"
Image: sha256:b7a2e0a9cb35c32cc9826d88e6caf8f695323845fe8920d4feea2d5f51b9b585 Container: d1e874f99879cb8ff6227b324aec63245224d5b57bf09f4b0572efd36249938d
Image: sha256:c088fb3db6149f8f9d865f07f86b556cb962f6e8d64334d84ebf1f262043e32f Container: 357aaa0d14d06605cb4953ffefa32f6b38668c3c343c05af6a1f360ecb45d23d
Image: sha256:611c01362fba46e6a4287c4779cfaaf90587984bccb92b5df5b05f55f8adaa32 Container: 98adaf58b002f32b40207efe230b494fdf63e9cc6ff2c823baec7055777daf7f
Image: sha256:89ea066aa4bfc02428b40dcee84afc4215b2a31accbf1b8d4ee3f86d61deeb5d Container: d4900ec54611d776b94adba2626e29c9aa356c2e6220731475d91140b930e1aa
```

Looks like this dangling image `611c01362fba` is used by a container and has a tag, but with `dangling=true` it is shown without 
a tag.

**Comment by stasadev** [comment]
With one more `ddev restart` a new dangling for ddev-dbserver appeared `c088fb3db614`, and one `d3c8a3de8610` for ddev-webserver:

```
$ docker images -f dangling=true
REPOSITORY   TAG       IMAGE ID       CREATED             SIZE
<none>       <none>    c088fb3db614   About an hour ago   670MB
<none>       <none>    d3c8a3de8610   2 months ago        1.16GB
<none>       <none>    611c01362fba   2 months ago        1.16GB
<none>       <none>    1444ab586442   2 months ago        119MB
```

```
$ docker container inspect $(docker container ls -aq) --format "Image: {{.Image}} Container: {{.Id}}"
Image: sha256:b7a2e0a9cb35c32cc9826d88e6caf8f695323845fe8920d4feea2d5f51b9b585 Container: 79b10388fb7365e5245401a4c9ec0c1b5ef5dc8305aaa1e632904f5385c4946c
Image: sha256:6381cee3e775c1d00f9981ef3a56d8685c8a69873454ec870783ccb467219bcc Container: 6221c2e742eae92c7feac5060b4e2c196837fd9a920f43bf49793f288c9fb606
Image: sha256:2623b36e9c1bc28217c9c16d0b4315c37a96a5d49d0d4dd9b95c59f5d7b7edce Container: a3e21b4de8237f16f516da59d2e24957282239ac761a748154e581a86f25b1df
Image: sha256:89ea066aa4bfc02428b40dcee84afc4215b2a31accbf1b8d4ee3f86d61deeb5d Container: d4900ec54611d776b94adba2626e29c9aa356c2e6220731475d91140b930e1aa
```

```
$ docker images | grep 1.1
drud/ddev-webserver                       v1.21.5-tryddevproject-5811-built          2623b36e9c1b   4 minutes ago       1.17GB
<none>                                    <none>                                     611c01362fba   2 months ago        1.16GB
<none>                                    <none>                                     d3c8a3de8610   2 months ago        1.16GB
drud/ddev-webserver                       v1.21.5                                    bcd7a520efce   2 months ago        1.16GB
```

This time, the used ddev-webserver was different `2623b36e9c1b` and not in dangling list.

>Maybe there's something not quite right there. Like the image statement... is it even useful?

I tried to find some kind of pattern, but something is wrong.

**Comment by rfay** [comment]
I wonder if it's this, 
* https://github.com/moby/moby/issues/45430

**Comment by rfay** [comment]
That was pulled *after* 23.0.5 was released, so should appear in 23.0.6
